pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -119,6 +121,98 @@ func (configgen *ConfigGeneratorImpl) BuildClusters(proxy *model.Proxy, push *mo
 	return resources, model.XdsLogDetails{AdditionalInfo: fmt.Sprintf("cached:%v/%v", cacheStats.hits, cacheStats.hits+cacheStats.miss)}
 }
 
+// BuildDeltaClusters will generate the deltas (add and delete) for a given proxy. Currently, only service changes are reflected with deltas.
+// Otherwise, we fall back onto generating everything.
+func (configgen *ConfigGeneratorImpl) BuildDeltaClusters(proxy *model.Proxy, push *model.PushContext,
+	updates *model.PushRequest, watched *model.WatchedResource) ([]*discovery.Resource, []string, model.XdsLogDetails) {
+	clusters := make([]*cluster.Cluster, 0)
+	resources := model.Resources{}
+	removedClusterNames := make([]string, 0)
+	services := make([]*model.Service, 0)
+	envoyFilterPatches := push.EnvoyFilters(proxy)
+	cb := NewClusterBuilder(proxy, push, configgen.Cache)
+	instances := proxy.ServiceInstances
+	// delta when only services are modified and we attempt deltas
+	// we can progressively optimize this
+	delta := updates != nil && allConfigKeysOfType(updates.ConfigsUpdated, gvk.ServiceEntry)
+	// if we can't use delta, fall back to generate all
+	if !delta || len(updates.ConfigsUpdated) == 0 {
+		cl, lg := configgen.BuildClusters(proxy, push)
+		return cl, make([]string, 0), lg
+	}
+	for s := range updates.ConfigsUpdated {
+		// get service that changed
+		service := proxy.SidecarScope.Service(host.Name(s.Name))
+		// SidecarScope.Service will return nil if the proxy doesn't care about the service OR it was deleted.
+		// we can just send that it was a removed_resource, because even if it wasn't, envoy will silently ignore.
```

Comment:
are the comments still valid, i can see you compare with watched resources below